source/Logger.cpp

```diff
@@ -15,31 +15,64 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
 #include "Logger.h"
 
+#include "text/Format.h"
+#include "GameVersion.h"
+
+#ifdef _WIN32
+#include "windows/WinVersion.h"
+#else
+#include <sys/utsname.h>
+#endif
+
 #include <iostream>
 #include <mutex>
 
 using namespace std;
 
 namespace {
-	function<void(const string &message)> logErrorCallback = nullptr;
-	mutex logErrorMutex;
+	function<void(const string &message, Logger::Level)> logCallback = nullptr;
+	mutex logMutex;
+}
+
+
+
+Logger::Session::Session()
+{
+	string message = "Logger session beginning. Game version: " + GameVersion::Running().ToString()
+		+ ". Detected operating system version: ";
+#ifdef _WIN32
+	message += WinVersion::ToString() + '.';
+#else
+	utsname uName;
+	uname(&uName);
+	message += string(uName.sysname) + ' ' + uName.release + ' ' + uName.version + '.';
+#endif
+	Log(message, Level::INFO);
+}
+
+
+
+Logger::Session::~Session()
+{
+	Log("Logger session end.", Level::INFO);
 }
 
 
 
-void Logger::SetLogErrorCallback(function<void(const string &message)> callback)
+void Logger::SetLogCallback(function<void(const string &message, Level)> callback)
 {
-	logErrorCallback = std::move(callback);
+	logCallback = std::move(callback);
 }
 
 
 
-void Logger::LogError(const string &message)
+void Logger::Log(const string &message, Level level)
 {
-	lock_guard<mutex> lock(logErrorMutex);
-	// Log by default to stderr.
-	cerr << message << endl;
+	lock_guard<mutex> lock(logMutex);
+	string formatted = Format::TimestampString(chrono::system_clock::now())
```

Comment:
I question the point of going off of the player's preferences for the log timestamps. Logs are likely something that will be shared around when someone is having a problem. I don't want to have to decipher three different timestamp formats.